definition_validator/variance.cc

```diff
@@ -0,0 +1,184 @@
+
+#include "common/typecase.h"
+#include "core/Symbols.h"
+#include "core/core.h"
+#include "core/errors/resolver.h"
+
+#include "variance.h"
+
+using namespace std;
+
+namespace sorbet::definition_validator::variance {
+
+class VarianceValidator {
+private:
+    const core::Loc loc;
+
+    VarianceValidator(const core::Loc loc) : loc(loc) {}
+
+    core::Variance invertVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return core::Variance::CoVariant;
+        }
+    }
+
+    string showVariance(const core::Variance variance) {
+        switch (variance) {
+            case core::Variance::CoVariant:
+                return ":out";
+            case core::Variance::Invariant:
+                return ":invariant";
+            case core::Variance::ContraVariant:
+                return ":in";
+        }
+    }
+
+    bool compatibleVariance(const core::Variance left, const core::Variance right) {
+        switch (left) {
+            case core::Variance::CoVariant:
+                return right != core::Variance::ContraVariant;
+            case core::Variance::Invariant:
+                return right == core::Variance::Invariant;
+            case core::Variance::ContraVariant:
+                return right != core::Variance::CoVariant;
+        }
+    }
```

Comment:
No strong preference because I think we're unlikely to add new enums to `core::Variance` ~ever, but I wonder if these would be better served as instance methods on `core::Variance` 🤔

Reply:
> add new enums to core::Variance 

😨

Reply:
We can't put methods on an `enum class`, should I just move these functions next to the type?

Reply:
ah I didn't know how enum classes worked. i'm fine with the way it currently is.